Additional C test infrastructure

pauli-oracle opened the pull request:
- [x] tests are added or updated

Very preliminary test infrastructure additions.

Rather than charging full steam ahead, I'd prefer some feedback to see if this is along desirable lines...

The intention is to provide a suite of functions along the lines of the existing strings_equal routine that test a condition and print a uniformly formatted message if it is false.

pauli-oracle commented:
@ekasper, is this along the lines of what you envisioned?

pauli-oracle commented:
Pretty much ignore the changes to `ssl_test.c` and `ssl_test_ctx_test.c`, these are fixing calls to `strings_equal()` so they compile.

The new functionality is in `testutil.{c,h}`.

I've updated my recent `stack_test.c` to use the new calls and to define a local suite of comparison functions.

levitte commented on test/testutil.c:
```diff
@@ -145,19 +145,110 @@ int run_tests(const char *test_prog_name)
 }
 
-int strings_equal(const char *desc, const char *s1, const char *s2)
+int strings_eq(const char *s1, const char *s2, const char *desc, ...)
```
A small nit; considering all other comparisons have singular names so far, it seems odd to have plural "strings" here...

pauli-oracle replied:
Yeah, I'd picked up on this already.  Should we use the singular or plural for the function names?  They are an argument to the macros so either is possible.

The original function was strings_equal() but that seemed unnatural to me.

levitte replied:
I prefer singular...  not sure how to rationalise it, though.  Maybe by saying that the name denotes the type, not the objects?

pauli-oracle replied:
Okay, I'll change them to singular tomorrow.

levitte commented on test/testutil.c:
```diff
@@ -145,19 +145,110 @@ int run_tests(const char *test_prog_name)
+    return 1;
+}
+
+int strings_ne(const char *s1, const char *s2, const char *desc, ...)
```
... and here

pauli-oracle commented on test/stack_test.c:
```diff
@@ -90,82 +92,67 @@ static int test_int_stack(void)
     for (i = 0; i < n; i++) {
-        if (sk_sint_num(s) != i) {
+        if (!int_eq(sk_sint_num(s), i, "int stack size %d", i)) {
             fprintf(stderr, "test int stack size %d\n", i);
```
Oops, I thought I'd replaced this already :(  It will go away.  @jon-oracle found it soon after I made the initial commit.

levitte commented on test/stack_test.c:
```diff
@@ -45,6 +45,8 @@ DEFINE_SPECIAL_STACK_OF_CONST(uchar, unsigned char)
 DEFINE_STACK_OF(SS)
 DEFINE_STACK_OF_CONST(SU)
 
+IMPLEMENT_COMPARISONS(char, char, "%c")
```
I'd like to see a static variant of `IMPLEMENT_COMPARISONS`...  it would be fitting here.

pauli-oracle replied:
I figured having some global implementations for common case comparisons would be useful.
However, there is no reason to not force each test program using these to declare exactly those they use themselves.  I'm kind of ambivalent here -- I'm happy to go either way and is seems the consensus is pushing for pertest instantiation.

pauli-oracle replied:
I guess the question is do we want all of these test functions defined in testutil.c and used everywhere or do we want them in testutil.h so tests can create individual instances of them?

At the moment there is the flexibility to do either (although local versions aren't static).  What is ideal here?

ekasper replied:
I think all the basic types should go to `testutil.h` (as they did). Can you give an example where tests would then still need individual instances? Are you thinking typedef'ed types?

ekasper commented on test/testutil.h:
```diff
@@ -141,10 +143,94 @@ __owur int run_tests(const char *test_prog_name);
+
+/*
+ * Check a boolean result for being true or false.
+ * They return 1 if the condition is true.
```
"... if the condition is true (i.e., the value is non-zero)".

Within OpenSSL, 0 sometimes means true :(

ekasper commented on test/testutil.h:
```diff
@@ -141,10 +143,94 @@ __owur int run_tests(const char *test_prog_name);
+ * the data type that was tested, |fmt| is a system provided printf format with
+ * arguments that spells out the failure details.
+ */
+#if defined(__GNUC__)
```
The `fail_message` method signature is hard to comprehend. Can you give an example in the comments how it is supposed to be used, and what the resulting output looks like?

If this  is not meant to be used by tests (but only by the other macros), then perhaps move into a separate header?

pauli-oracle replied:
I'm assuming that the output of fail_message is in a dreadful format and will be changed.  I'll add better descriptions as to what its arguments are and do.

ekasper commented on test/testutil.c:
```diff
@@ -145,19 +145,110 @@ int run_tests(const char *test_prog_name)
     return EXIT_SUCCESS;
 }
 
+void fail_message(const char *desc, va_list args, const char *type,
```
Ideally I'd like to see `__FILE__` and `__LINE__` here. You'd need to move the implementation to a header, though.

With `__FILE__` and `__LINE__` we probably won't even need `desc` in most cases, so we could make custom descriptions optional (less typing):

`ret = TEST_int_eq(a, b);`

And optionally

```
ret = TEST_int_eq_msg(a[i], b[i], "%d th element mismatch", i);
```

pauli-oracle replied:
I don't see how adding FILE & LINE to the arguments is going to be possible.  We'd want macros to do this which then call the underlying functions with the extra arguments.  How do I generate the macros automatically?  Generating the functions is easy enough and doing macros for predefined ones would be possible (by listing them out).  How would I go with e.g. IMPLEMENT_COMPARISONS(struct fred *, fredptrs, "%p")
I'd need to generate fredpts_eq(...) and the wrapper macro TEST_fredpts_eq(...).

pauli-oracle replied:
Is there even an OpenSSL approved method of varadic macros?

levitte replied:
Not as far as I know. That came with later C standards...

pauli-oracle replied:
It would be possible to have macro versions of these without the description or at least without the description with arguments.  That might be sufficient.  I'll mulled it over a bit....

ekasper commented on test/testutil.h:
```diff
@@ -141,10 +143,94 @@ __owur int run_tests(const char *test_prog_name);
+    DEFINE_COMPARISONS(type, name, fmt)
+
+/*
+ * Define eight comparisons between pairs of integers and unsigned integers.
```
It'd be nice if you could spell out, at least once here, what the resulting method signatures are.

ekasper commented on test/ssl_test.c:
```diff
@@ -160,25 +160,25 @@ static int check_compression(HANDSHAKE_RESULT *result, SSL_TEST_CTX *test_ctx)
-    ret &= strings_equal("ExpectedNPNProtocol",
-                         test_ctx->expected_npn_protocol,
-                         result->client_npn_negotiated);
+    ret &= strings_eq(result->client_npn_negotiated,
```
Here's an idea for a future PR: collect all errors in a static variable and make each test automatically fail if one of the test assertions fails.

Not sure how ugly the implementation would be with the tooling we have (C90) but worth a thought.

pauli-oracle replied:
Collecting the errors wouldn't be too hard.  A stack of strings, malloc buffers and sprintf into them.
Fail if the stack isn't empty.

ekasper commented on test/testutil.h:
```diff
@@ -141,10 +143,94 @@ __owur int run_tests(const char *test_prog_name);
+ * Otherwise, they return 0 and pretty-prints diagnostics using |desc|.
  */
-int strings_equal(const char *desc, const char *s1, const char *s2);
+int bool_true(int, const char *desc, ...);
```
And on pointers specifically, I'd probs call this `TEST_ptr` - less typing, and I think `if (!TEST_ptr(p))` is easier to read than the double negative `if (!TEST_ptr_not_null(p))`.

We may also want an aborting flavour of this one. It's bad value for effort to write all the extra `goto end` boilerplate to exit cleanly on say malloc failures in tests (which just never happen). You don't have to do it in this PR, though.

levitte replied:
@ekasper, how would you rename `ptr_null`?

ekasper replied:
Does this need its own macro? You could simply use `TEST_ptr_eq(p, NULL)`.

levitte replied:
Ah, so you were thinking `TEST_ptr` *in addition* to `TEST_ptr_eq` and `TEST_ptr_ne`? That escaped me, and in that case I have no concerns in the matter...

pauli-oracle replied:
I've got some embryonic ideas as to how to do automated malloc testing -- i.e. run through all the calls to malloc and cause each to fail in turn.  Having test cases that abort early due to this would be detrimental to the effort.

Are we agreed to prefix these functions with TEST_ and to shorten the ptr not null name?

pauli-oracle commented:
Currently not done: FILE & LINE support via macros.
Prefixing functions with TEST_.
I think I've addressed the other concerns.

pauli-oracle commented:
How's this?  I've added TEST_ macros that use the description field to display the file and line number -- the user doesn't get to provide a description here but they can always fall back to the underlying function directly for this.

I've updated some of the tests in `stack_test.c` to use the macros.

richsalz commented:
we have the tooling to do malloc failures; look at the code in crypto/mem.c

richsalz commented:
No varadic macro support.

pauli-oracle commented:
I've seen the crypto mem code, it is sufficient.  I'm still figuring out a good way to attack the problem so such testing can become integral to the test framework.  There's little point doing ad-hoc tests, each implemented differently.  It's a WIP but I'm reasonably confident it's workable.  But that is for later.

ekasper commented on test/testutil.h:
```diff
@@ -141,10 +143,201 @@ __owur int run_tests(const char *test_prog_name);
+ * The following macros provide wrapper calls to the test functions with
+ * a default description that indicates the file and line number of the error.
+ */
+#define TEST_int_eq(a, b)   test_int_eq(a, b, "%s: %d", __FILE__, __LINE__)
```
This looks good! However I'd still like to preserve file and line numbers for the custom error messages too. Now we can't do that in general (because no support for variadic macros in C90) but you could do something like this I think?

```
# if !defined(__STDC_VERSION__) || __STDC_VERSION__ < 199901L
# define TEST_error test_error_c90
#else
# define TEST_error(...) test_error(__FILE__, __LINE__, __VA_ARGS__)
#endif
```

where `test_error_c90` takes a format string and args, and `test_error` additionally takes the file and line number.

Then hide the lowercase functions in a separate internal header (`testutil_internal.h`) so that tests don't use them directly. Code that previously called `test_int_eq` would do

```
if (!TEST_int_eq(a[i], b[i])) {
  TEST_error("%d th element mismatch", i);
}
```

Or you could also provide a macro per comparison, `TEST_int_eq_msg(a, b, msg)`, etc.

What do you think?

pauli-oracle replied:
I'm not convinced that `TEST_error` needs to show the file and line number.  With the usage example given, the file and line will be produced by `TEST_int_eq` and would be redundant.  Perhaps call the function `TEST_extra` or `TEST_info`

I don't like the `TEST_int_eq_msg` suggestion because we don't have variadic macros so there can be no arguments to the message.

Are you suggesting hiding the test_xxx functions in an header but still creating them in testutil.c ?

ekasper replied:
I meant you'd do `TEST_int_eq_msg`same as `TEST_error` - variadic macros where you can, and  fall back to a function (losing the file and line number) otherwise. I _think_ that works?

Re file and line numbers, I don't think they ever hurt. I'd probably have `TEST_error` for errors (starts with ERROR or FAIL as you have it) and `TEST_info` for info (starts with INFO) to make it easy to grep for failure lines - but I don't really have a strong opinion. Whatever feels most natural to you!

Re test_xxx functions, I think so? It's not so important - I only suggested moving the declarations to another header to make it easier to find the public bits you're supposed to use in tests.

pauli-oracle replied:
I think the TEST_xxx_msg flavours can be made to work, it will introduce quite a bit of bloat to the header file -- is the gain worthwhile?  The `if (TEST_xxx()) TEST_ERROR()` sequence seems reasonable and sufficiently terse.  At least to me.  I kind of suspect that the TEST_xxx__msg versions won't be heavily used -- people often seem to not like writing tests and when they do, they make them as short and unhelpful as possible.  Slightly exaggerated perhaps.

You're right, the line numbers won't hurt.  I'll add the info versions.

The declarations need to be visible in testutil.h to avoid compiler warnings about undefined functions.  Whether they are in another .h file or not seems to be the concern.  I'd don't have a strong feeling either way here.

ekasper commented on test/testutil.h:
```diff
@@ -141,10 +143,201 @@ __owur int run_tests(const char *test_prog_name);
+#define TEST_uchar_gt(a, b) test_uchar_gt(a, b, "%s: %d", __FILE__, __LINE__)
+#define TEST_uchar_ge(a, b) test_uchar_ge(a, b, "%s: %d", __FILE__, __LINE__)
+
+#define TEST_lint_eq(a, b)  test_lint_eq(a, b, "%s: %d", __FILE__, __LINE__)
```
I'd suggest `long` and `ulong` ("lint" makes me think of linters).

ekasper commented on test/testutil.c:
```diff
@@ -145,19 +145,114 @@ int run_tests(const char *test_prog_name)
+DEFINE_COMPARISONS(unsigned int, uint, "%u")
+DEFINE_COMPARISONS(char, char, "%c")
+DEFINE_COMPARISONS(unsigned char, uchar, "%u")
+DEFINE_COMPARISONS(long int, lint, "%ld")
```
Simply `long` and `unsigned long` (drop the `int`).

pauli-oracle commented:
I think this addresses the next round of changes.  I've added the TEST_error macro and hidden the definition macros (in testutil.c rather than a separate header file).

pauli-oracle commented:
Regarding the TEST_xxx_msg flavours, I can't help but feel that they aren't worthwhile.
The whole point of these tests is that they should pass and the message is never seen.  If a test does fail, somebody is going to have to go and debug the code and the file + line number is adequate to get started.  A message is a nicety but isn't really necessary.  Plus messages are available with test_ERROR and test_INFO and can easily be added to a failing test.

pauli-oracle commented:
I've added the TEST_xxx_msg flavours.  See if you feel the extra complexity is worthwhile or not.

pauli-oracle commented:
Reminder to self: don't commit code last thing on a Friday :(

ekasper commented:
Agreed, I don't particularly feel the `_msg` macros are worth it either. I was just throwing out a couple of alternatives but since you feel the same way, let's toss them. Let's not add bloat for things we're not sure we need.

pauli-oracle commented:
Cool, I'll fix things on Monday.

pauli-oracle commented:
I've tossed out the _msg flavours.  I've converted some additional test programs to use this message framework.

ekasper commented on test/asn1_internal_test.c:
```diff
@@ -70,7 +70,7 @@ static int test_standard_methods()
     for (tmp = standard_methods, i = 0; i < OSSL_NELEM(standard_methods);
          i++, tmp++) {
-        if ((*tmp)->pkey_id < last_pkey_id) {
+        if (!TEST_int_ge((*tmp)->pkey_id, last_pkey_id)) {
```
I'm not quite sure what this test does but it doesn't seem to be exercising a failure here, so I think this should stay as is.

pauli-oracle replied:
It is a failure condition of sorts.  Setting `last_pkey_id` to zero skips the normal exit and prints the method table.  It is up to the user to find the one that is sorted incorrectly.  Adding the message pinpoints this directly.

I'm still happy to revert this change but I think it adds some value.

pauli-oracle replied:
Reworked this to keep the original condition and use TEST_error instead.

ekasper replied:
Ah, gotcha! I was reading the error condition backwards, my bad.

ekasper commented on test/cipherlist_test.c:
```diff
@@ -136,18 +136,18 @@ static int test_default_cipherlist(SSL_CTX *ctx)
-    if (num_ciphers != num_expected_ciphers) {
-        fprintf(stderr, "Expected %d supported ciphers, got %d.\n",
-                num_expected_ciphers, num_ciphers);
+    if (!TEST_int_eq(num_ciphers, num_expected_ciphers)) {
```
Looking at these, I had another thought. Can we define the macros as

```
#define TEST_int_eq(a, b) test_int_eq(__FILE__, __LINE__, #a, #b, a, b)
```

i.e., stringify the value(s) under test and print upon failure. Then the error line contains enough information that you could mostly scrap the `TEST_info` lines.

levitte replied:
That makes for a very different output, no?  Stringifying `a` and `b` doesn't display any *values*, which I see as valuable information.

levitte replied:
(never mind the comment I just make...  deleted)

pauli-oracle replied:
Done.

pauli-oracle replied:
And this simplified things a deal too.

pauli-oracle replied:
I've left the `TEST_info` lines for the moment.

ekasper commented on test/testutil.c:
```diff
@@ -145,18 +145,261 @@ int run_tests(const char *test_prog_name)
+}
+
+
+int test_mem_eq(const void *s1, const void *s2, size_t n, const char *desc, ...)
```
You could consider doing `test_mem_eq(const void *s1, size_t s1_len, const void *s2, size_t s2_len)`, then tests don't have to check the length separately.

pauli-oracle replied:
Changed.

ekasper commented on test/testutil.c:
```diff
@@ -145,18 +145,261 @@ int run_tests(const char *test_prog_name)
+      return 1;
+    if (s1 == NULL || s2 == NULL || memcmp(s1, s2, n) != 0) {
+        va_start(ap, desc);
+        test_fail_message(1, desc, ap, "memory", "%p == %p [%u]", s1, s2, (unsigned)n);
```
It's probably more useful to print out the buffer contents here?

pauli-oracle replied:
Done -- with a limit.

ekasper commented on test/testutil.h:
```diff
@@ -140,18 +142,169 @@ __owur int run_tests(const char *test_prog_name);
 /*
- * Returns 1 if |s1| and |s2| are both NULL or equal.
- * Otherwise, returns 0 and pretty-prints diagnostics using |desc|.
+ * Ppretty print a failure message.
```
nit, typo: Ppretty -> Pretty

ekasper commented:
I have new comments - but in my defense you have added new functionality :) In any case, these are mostly optional suggestions, feel free to leave some of these ideas for later.

pauli-oracle commented:
I've added more new functionality :)
This is coming together nicely I think.
I've also added a test case to test this test functionality.  It will produce spurious error output but won't fail unless something actually goes wrong.

pauli-oracle commented:
I've cleaned out some of the more redundant info messages from the updated tests.

ekasper commented on test/testutil.c:
```diff
@@ -145,18 +148,284 @@ int run_tests(const char *test_prog_name)
+    
+    for (i=0; i<2*n; i++) {
+        unsigned char c = (i & 1) != 0 ? p[i / 2] & 15 : p[i / 2] >> 4;
+        out[i] = c["0123456789abcdef"];
```
Is this c90?

levitte replied:
That's C since way back.  A little quirk of the language.  Ever tried this?

``` C
#include <stdio.h>

int main()
{
    printf("%c", 3["123456789"]);
}
```

Of course, the other way around, `"123456789"[3]`, is just as possible and equivalent.

richsalz replied:
yeah, let's flip it around to avoid questions.

pauli-oracle replied:
This originally comes from K&R C and has remained ever since.  `a[b]` is defined to be `*(a + b)`.  Thus, the order doesn't matter.

Little quirk in my coding to hex conversions...

ekasper commented on test/cipherlist_test.c:
```diff
@@ -136,18 +136,15 @@ static int test_default_cipherlist(SSL_CTX *ctx)
-            fprintf(stderr, "Wrong cipher at position %d: expected %x, "
-                    "got %x\n", i, expected_cipher_id, cipher_id);
+        if (!TEST_int_eq(cipher_id, expected_cipher_id)) {
+            TEST_info("Wrong cipher at position %d: expected %x, "
```
`TEST_error`? (Everywhere where errors are logged?)

pauli-oracle replied:
The TEST_int_eq will produce the error message.  This is extra additional info: the position where the error occurred.  I don't think it should raise a second error, since it isn't.  I'll get rid of the duplicated information though.  There are probably a few cases like this still.

ekasper commented on test/test_test.c:
```diff
@@ -0,0 +1,268 @@
+ */
+
+/*
+ * Copyright (c) 2017 Oracle and/or its affiliates.  All rights reserved.
```
I believe we avoid custom copyright statements in file headers these days. (The CLA does not transfer copyright - just that additional copyrights don't need to be listed in our headers.)

pauli-oracle commented on test/testutil.c:
```diff
@@ -145,18 +148,284 @@ int run_tests(const char *test_prog_name)
+            PRINTF_FORMAT(5, 6);
+
+static void test_fail_message_va(int error, const char *file, int line,
+                              const char *type, const char *fmt, va_list ap)
```
Good suggestion :)

pauli-oracle commented:
Should be good to go.  Run `make test TESTS=test_test V=1` to see the output of failed messages.  All the information is there but if anyone feels changes are desirable, we should do them before any announcement.

I've also updated test/README to not include the bit about editing the makefile and to have a short section about this.

pauli-oracle commented:
Any thoughts about dedicating a code health Tuesday to modifying existing test cases to use this infrastructure more uniformly?

richsalz commented:
@pauli-oracle that's a great idea!  I'll push it.  Looks like @levitte;s conerns were address.  I also approve and can  merge.

pauli-oracle commented:
Thanks @richsalz.  The single build failure is due to a problem with documentation (i.e. not part of this PR).

richsalz commented:
I needed these changes:

```
diff --git a/test/test_test.c b/test/test_test.c
index 842b449..de39917 100644
--- a/test/test_test.c
+++ b/test/test_test.c
@@ -254,8 +254,9 @@ err:

 static int test_messages(void)
 {
-    TEST_INFO("This is an %s message.", "info");
-    TEST_ERROR("This is an %s message.", "error");
+    TEST_info("This is an %s message.", "info");
+    TEST_error("This is an %s message.", "error");
+    return 0;
 }

```

And also the test_test recipe needs to be changed.  Not sure how to say "failure is success"  @levitte ?

pauli-oracle commented:
Sorry, I hadn't pushed that capitalisation change yet.  I realised I'd not executed those two functions in the test.  Done now.

There is no need to modify the recipe, the `test_test.c` program checks the _failing_ return codes and acts accordingly.  The failure messages will be printed out but the executable returns the correct code and the test passes.

pauli-oracle commented:
This could be a pain when grepping for ERRORs.

I also just noticed that there is no test name in the diagnostics...

richsalz commented:
TEST_INFO is a struct, as you saw I assume.Do you want to push a change to fix the logging issue?

richsalz commented:
@levitte or @ekasper reconfirm?

pauli-oracle commented:
I've fixed the test_test.c issue.  I'm not concerned about the test name -- we've got file/line number information in the output already so it isn't essential.  The error messages are a little busy already.

richsalz commented:
So we just need a second reviewer for that last bit.

pauli-oracle commented:
Both lines of it :)

t-j-h commented:
Which two lines?

pauli-oracle commented:
test/test_test.c lines 263 & 264.  Well, actually that whole five line function and the line at the bottom that adds it as a test.  But only two lines of content :)

pauli-oracle commented:
```
static int test_messages(void)
{
    TEST_info("This is an %s message.", "info");
    TEST_error("This is an %s message.", "error");
    return 1;
}
```
and 

```
    ADD_TEST(test_messages);
```

t-j-h commented:
+1 on the "last bit".

richsalz commented:
merged.  thanks for this, and the reviews folks!

pauli-oracle commented:
The other code health thought I've had is "fix an example day".  I.e. go through the examples and get them working and bring them up to date or write new examples.
